src/prpy/planning/base.py

```diff
@@ -52,13 +52,27 @@ def __init__(self, func):
 
     def __call__(self, instance, robot, *args, **kw_args):
         env = robot.GetEnv()
-
-        with Clone(env, clone_env=instance.env):
-            planning_traj = self.func(instance, Cloned(robot), *args, **kw_args)
-            traj = openravepy.RaveCreateTrajectory(env, planning_traj.GetXMLId())
-            traj.Clone(planning_traj, 0)
-
-        return traj
+        defer = kw_args.get('defer')
+
+        with Clone(env, clone_env=instance.env, lock=True, unlock=False) \
+                as cloned_env:
+
+            def call_planner():
+                planning_traj = self.func(instance, Cloned(robot),
+                                          *args, **kw_args)
+                traj = openravepy.RaveCreateTrajectory(
+                    env, planning_traj.GetXMLId()
+                )
+                traj.Clone(planning_traj, 0)
+                cloned_env.Unlock()
```

Comment:
The planning environment may not be unlocked if `self.func` raises an exception. The `Unlock()` call needs to be in a `finally`-block. Or, alternative, you can use `contextlib` to manage the `Clone()` context manager.